src/reader_disc.cpp

```diff
@@ -0,0 +1,559 @@
+      default:
+        printf("Unknown DVD protection shceme!\n");
+        break;
+      }
+      printf("Disc regions       : ");
+      if (dev->media.dvdcss.regmask != 0xFF) {
+        for (int i = 0; i < 8; i++)
+          if (!((dev->media.dvdcss.regmask >> i) & 1)) {
+            printf("%d", i + 1); /*dev->rpc.region = i+1;*/
+          }
+        printf("\n");
+      } else {
+        printf("Invalid region mask!\n");
+      }
+    }
+
+    switch (dev->media.dvdcss.protection) {
+    case 0: // unprotected DVD
+      dev->media.dvdcss.method = DVDCSS_METHOD_NONE;
+      break;
+    case 1: // CSS/CPPM protected DVD
+      // just to auth to be able read data
+      //				scnt=1;
+      dev->media.dvdcss.method = DVDCSS_METHOD_KEY;
+      // dev->media.dvdcss.method = DVDCSS_METHOD_DISC;
+      if (css_disckey(dev)) {
+        printf("DVD auth failure!\n");
+        goto exit_reader;
+      }
+      initAllCSSKeys(dev);
+      break;
+    case 2: // CPRM - protected DVD
+      // just to auth to be able read data
+      dev->media.dvdcss.method = DVDCSS_METHOD_KEY;
+      if (css_disckey(dev)) {
+        printf("DVD auth failure!\n");
+        goto exit_reader;
+      }
+      break;
+    default:
+      //	printf("Unknown DVD protection scheme:
+      //%02X\n",dev->media.dvdcss.protection);
+      dev->media.dvdcss.method = DVDCSS_METHOD_NONE;
+      break;
+    }
+  } else {
+    printf("Unsupported media!\n");
+    goto exit_reader;
+  }
+
+  if (parm->pass >= PASS_RECOVER)
+    scnt = 1;
+
+  map->lock();
+  if (parm->pass < PASS_RECOVER) {
+    map->get_next(&lba, BM_WAIT, &scnt);
+  } else {
+    bool rce, wce;
+    printf("Recover pass - trying to disable read cache...\n");
+    get_cache(dev, &rce, &wce);
+    if (!rce) {
+      printf("Read cache already disabled:)\n");
+    } else {
+      if (set_cache(dev, 0, wce)) {
+        printf("It seems drive does not support disabling read cache!\n");
+      } else {
+        printf("Read cache disabled successfully\n");
+      }
+    }
+    map->get_next(&lba, BM_FAIL, &scnt);
+  }
+  map->set(lba, BM_READ, scnt);
+  map->unlock();
+
+  if (dev->media.dvdcss.protection == 0x01) {
+    for (int i = 0; i < DVD_KEY_SIZE; i++)
+      dev->media.dvdcss.TK[i] = 0;
+    p_title = dev->media.dvdcss.p_titles;
+    descramble = 0;
+    if (!p_title) {
+      title_next = 0xFFFFFFFF;
+    } else {
+      title_next = p_title->i_startlb;
+    }
+    printf("\nSetting title key: %02X:%02X:%02X:%02X:%02X  for %x-%x, CSS=%d\n",
+           dev->media.dvdcss.TK[0], dev->media.dvdcss.TK[1],
+           dev->media.dvdcss.TK[2], dev->media.dvdcss.TK[3],
+           dev->media.dvdcss.TK[4], 0, title_next, descramble);
+  }
+
+  gettimeofday(&t_e, NULL);
+  seek(dev, 0);
+  while (lba < dev->media.capacity) {
+    if (lba == 0xFFFFFFFF)
+      goto exit_reader;
+    t_s.tv_sec = t_e.tv_sec;
+    t_s.tv_usec = t_e.tv_usec;
+
+    if (dev->media.type & DISC_CD) {
+      read(dev, dev->rd_buf, lba, scnt);
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)